homeassistant/helpers/event.py

```diff
@@ -1133,16 +1132,36 @@ def async_track_point_in_utc_time(
     # having to figure out how to call the action every time its called.
     job = action if isinstance(action, HassJob) else HassJob(action)
 
-    cancel_callback = hass.loop.call_at(
-        hass.loop.time() + point_in_time.timestamp() - time.time(),
-        hass.async_run_hass_job,
-        job,
-        utc_point_in_time,
-    )
+    cancel_callback: Optional[asyncio.TimerHandle] = None
+
+    @callback
+    def run_action() -> None:
+        """Call the action."""
+        nonlocal cancel_callback
+
+        now = time_tracker_utcnow()
+
+        # Depending on the available clock support (including timer hardware
+        # and the OS kernel) it can happen that we fire a little bit too early
+        # as measured by utcnow(). That is bad when callbacks have assumptions
+        # about the current time. Thus, we rearm the timer for the remaining
+        # time.
+        delta = (utc_point_in_time - now).total_seconds()
```

Comment:
Since call_later also accepts a float, should we pass whole seconds here?

Reply:
Slightly faster could be to use `utc_point_in_time.timestamp() - time.time()` here.

Reply:
We cannot use `time.time()` here because `async_fire_time_changed()` (in `tests/common.py`) does not patch it so we do not get the advanced time. Doing so sends me down the path with other tests breaking that I had to revert. So I would like to leave this as is since I have already struggled so much with this PR, starting out as a one-line fix :-)

I also think mixing two time packages hinders readability a bit so I wonder if it is really preferable without identified performance issues. Again, this is executed once when a timer expires so we are saving at most a few μs per second.

Reply:
About whole seconds, I do not understand. Which part is not a float?

Reply:
I meant: do we want to wait a whole second if we are only .5 seconds away from the target time.

Reply:
We don't and we won't; `total_seconds()` returns 0.5 in that case.